sql/core/src/main/scala/org/apache/spark/sql/execution/ExplainUtils.scala

```diff
@@ -0,0 +1,237 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.expressions.{Expression, PlanExpression}
+import org.apache.spark.sql.catalyst.plans.QueryPlan
+import org.apache.spark.sql.catalyst.trees.TreeNodeTag
+
+object ExplainUtils {
+  /**
+   * Given a input physical plan, performs the following tasks.
+   *   1. Computes the operator id for current operator and records it in the operaror
+   *      by setting a tag.
+   *   2. Computes the whole stage codegen id for current operator and records it in the
+   *      operator by setting a tag.
+   *   3. Generate the two part explain output for this plan.
+   *      1. First part explains the operator tree with each operator tagged with an unique
+   *         identifier.
+   *      2. Second part explans each operator in a verbose manner.
+   *
+   * Note : This function skips over subqueries. They are handled by its caller.
+   *
+   * @param plan Input query plan to process
+   * @param append function used to append the explain output
+   * @param startOperationID The start value of operation id. The subsequent operations will
+   *                         be assigned higher value.
+   *
+   * @return The last generated operation id for this input plan. This is to ensure we
+   *         always assign incrementing unique id to each operator.
+   *
+   */
+  private def processPlanSkippingSubqueries[T <: QueryPlan[T]](
+      plan: => QueryPlan[T],
+      append: String => Unit,
+      startOperatorID: Int): Int = {
+
+    // ReusedSubqueryExecs are skipped over
```

Comment:
This is too implicit. How about
1. `getSubqueries` collects `BaseSubqueryExec` instead of `BaseSubqueryExec#child`
2. `processPlan` checks the collected `BaseSubqueryExec`, if it's `ReusedSubqueryExec` then skip calling `processPlanSkippingSubqueries`

Reply:
@cloud-fan Actually we do collect BaseSubqueryExec#child. ReusedSubqueryExec's child is BasedSubqueryExec.. So i will put this instead :
1. getSubqueries collects the child plan BaseSubqueryExec. 
2. processPlan checks the collected child plan of BaseSubqueryExec. If child plan is an instance of BaseSubqueryExec (happens in case ReusedSubqueryExec), then it skips calling `processPlanSkippingSubqueries` in order to avoid printing the same subquery plan more than once.

Reply:
> Actually we do collect BaseSubqueryExec#child

my suggestion was to collect `BaseSubqueryExec`, so that we can check if it's a `ReusedSubqueryExec` in `processPlan`

Reply:
@cloud-fan oh.. sorry.. thought you just wanted the comment changed. Let me see what i need to change further.. i remember i started with that and got into some issues.. and ended up with current code..

Reply:
@cloud-fan then i think in `processPlanSkippingSubqueries` i would have to have something in the start like :
```
if (plan.isInstanceOf[ReusedSubqueryExec]) {
  return return startOperatorID
}
val curPlan = if (plan.isInstanceOf[SubqueryExec]) {
   plan.asInstanceOf[SubqueryExec].child
} else {
   plan
}
```
Actually i wanted to avoid the 2nd if block.. thats why i had it this way. Kindly let me know what you think.

Reply:
@cloud-fan Thanks for your suggestion. I have made the change per ur input.